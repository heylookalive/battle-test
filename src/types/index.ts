export enum GameState {
  ACTIVE = 'ACTIVE',
  OVER = 'OVER',
}

export enum PlayerTypes {
  HERO = 'HERO',
  MONSTER = 'MONSTER',
}

export type PlayerState = {
  type: PlayerTypes;
  health: number;
  dice?: number[];
};

// Lots of ways to skin a cat here, could have a "players" object to allow for
// more than two defined players.
export type BattleState = {
  players: {
    playerOne: PlayerState;
    playerTwo: PlayerState;
  };
  message: string;
  gameState: GameState;
};

export type BattleActions = { type: 'attack' };
