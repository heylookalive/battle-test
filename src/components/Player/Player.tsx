import React from 'react';
import styled from 'styled-components';
import { PlayerState, PlayerTypes } from '../../types';
import Dice from '../Dice/Dice';
import HealthBar from '../HealthBar/HealthBar';
import heroImage from './images/hero.jpg';
import monsterImage from './images/monster.jpg';

// Define own prop Type which can be altered in the future if we want.
export type PlayerProps = PlayerState;

const PlayerDetail = styled.div`
  display: grid;
  grid-template-columns: 1fr 30px 160px;
  min-height: 296px;
`;

const PlayerImage = styled.img`
  display: block;
  max-width: 100%;
`;

export default function Player({ type, health, dice }: PlayerProps) {
  const playerLabel = type === PlayerTypes.HERO ? 'Player' : 'Monster';

  return (
    <div>
      <PlayerDetail>
        <div>
          <PlayerImage
            src={type === PlayerTypes.HERO ? heroImage : monsterImage}
            alt={`${playerLabel} image`}
          />
        </div>
        <HealthBar value={health} />
        <div>
          {dice && (
            <>
              <Dice value={dice[0]} />
              <Dice value={dice[1]} />
            </>
          )}
        </div>
      </PlayerDetail>
      <h3>{playerLabel}</h3>
    </div>
  );
}
