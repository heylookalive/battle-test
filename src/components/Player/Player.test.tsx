import { render } from '@testing-library/react';
import React from 'react';
import { PlayerTypes } from '../../types';
import Player, { PlayerProps } from './Player';

describe('<Player />', () => {
  it('renders hero as expected', () => {
    const hero: PlayerProps = {
      type: PlayerTypes.HERO,
      health: 100,
      dice: [1, 2],
    };

    const { container } = render(<Player {...hero} />);
    expect(container).toMatchSnapshot();
  });

  it('renders monster as expected', () => {
    const monster: PlayerProps = {
      type: PlayerTypes.MONSTER,
      health: 72,
      dice: [5, 6],
    };

    const { container } = render(<Player {...monster} />);
    expect(container).toMatchSnapshot();
  });
});
