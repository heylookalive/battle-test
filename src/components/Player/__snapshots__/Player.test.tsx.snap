// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> renders hero as expected 1`] = `
<div>
  <div>
    <div
      class="sc-AxhUy bIlCyN"
    >
      <div>
        <img
          alt="Player image"
          class="sc-AxgMl hFZtTr"
          src="hero.jpg"
        />
      </div>
      <div
        class="sc-AxjAm ikqlAb"
      >
        <div
          class="sc-AxirZ dQUJrW"
        >
          <div
            class="sc-AxiKw kQyfxs"
            value="100"
          />
        </div>
        <div
          class="sc-AxhCb cinyOp"
          data-testid="healthbar-label"
        >
          100
        </div>
      </div>
      <div>
        <div
          class="face"
        >
          <span
            class="pip"
          />
        </div>
        <div
          class="face"
        >
          <span
            class="pip"
          />
          <span
            class="pip"
          />
        </div>
      </div>
    </div>
    <h3>
      Player
    </h3>
  </div>
</div>
`;

exports[`<Player /> renders monster as expected 1`] = `
<div>
  <div>
    <div
      class="sc-AxhUy bIlCyN"
    >
      <div>
        <img
          alt="Monster image"
          class="sc-AxgMl hFZtTr"
          src="monster.jpg"
        />
      </div>
      <div
        class="sc-AxjAm ikqlAb"
      >
        <div
          class="sc-AxirZ dQUJrW"
        >
          <div
            class="sc-AxiKw gsrgqA"
            value="72"
          />
        </div>
        <div
          class="sc-AxhCb cinyOp"
          data-testid="healthbar-label"
        >
          72
        </div>
      </div>
      <div>
        <div
          class="face"
        >
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
        </div>
        <div
          class="face"
        >
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
          <span
            class="pip"
          />
        </div>
      </div>
    </div>
    <h3>
      Monster
    </h3>
  </div>
</div>
`;
